src/style-spec/expression/definitions/in.js

```diff
@@ -55,37 +73,59 @@ class In implements Expression {
             return context.error(`Expected first argument to be of type boolean, string, number or null, but found ${toString(needle.type)} instead`);
         }
 
-        return new In(needle, haystack);
+        if (args.length === 4) {
+            const fromIndex = context.parse(args[3], 3, NumberType);
+            if (!fromIndex) return null;
+            return new In(type, needle, haystack, fromIndex);
+        } else {
+            return new In(type, needle, haystack);
+        }
     }
 
     evaluate(ctx: EvaluationContext) {
         const needle = (this.needle.evaluate(ctx): any);
         const haystack = (this.haystack.evaluate(ctx): any);
 
-        if (needle == null || !haystack) return false;
+        if (!haystack && this.type.kind === 'boolean') return false;
```

Comment:
Previously `in` would allow the needle to be null but would always return false, this means that if you had an array with null values in it and you were searching for the nulls it would always return false.

```javascript
[1,2,3,null,5].indexOf(null) >= 0 === true
```

In addition, the check returns false if the haystack is null. This was raised in [the initial review](https://github.com/mapbox/mapbox-gl-js/pull/8876/files#r340824010) for the `in` expression.

> The difference is in whether or not we're asserting that the value is a string or array. If we do ['string', ['get', 'str']] and str is null, then we get a type error. But if we don't assert it's a string or array and just directly return null then the check on line 65 of in.js returns false because there's no value that can be checked

I've left it so that the `in` expression keeps this return path, but potentially we need to make `index-of` return -1 so it's consistent between the two expressions?

Reply:
We need to keep the type checking (e.g. `['string', ['get', 'str']] // -> null` should return a type error). But maybe we should allow the needle to be null. See https://github.com/mapbox/mapbox-gl-js/pull/7197#pullrequestreview-153154765

>since "in" is using the same equality semantics as "==" (i.e. JS ===), it should have the same constraints, so it can't be used with types where === doesn't make sense. Specifically the type of needle should be limited to these types. If the static type is not one of those five, it should be a parse error. Or, if the static type is value and the runtime type isn't one of the other four, it should be an evaluation error.

`null === null` returns `true` though so it seems like we should accept that value and this was an oversight. Technically, `undefined` should work too, but I'm not sure we want to do that. You would need to update the `isComparableRuntimeValue` function to include `null` and update the error on line 68 to mention that `null` is an acceptable value

Reply:
Just letting you know that this was updated in the latest commit and I added a test case as well